[MachineLateInstrsCleanup] Handle multiple kills for a preceding definition.

JonPsson1 opened the pull request:
When removing a redundant definition in order to reuse an earlier identical one it is necessary to remove any earlier kill flag as well.

Previously, the assumption has been that any register that kills the defined Reg is enough to handle for this purpose, but this is actually not quite enough. A kill of a super-register does not necessarily imply that all of its subregs (including Reg) is defined at that point: a partial definition of a register is legal. This means Reg may have been killed earlier and is not live at that point.

This patch changes the tracking of kill flags to allow for multiple flags to be removed: instead of remembering just the single / latest kill flag, a vector is now used to track and remove them all. TinyPtrVector seems ideal for this as there are only very rarely more than one kill flag, and it doesn't seem to give much difference in compile time.

The kill flags handling here is making this pass much more complicated than it would have to be. This pass does not depend on kill flags for its own use, so an interesting alternative to all this handling would be in case there is no real use of them at this late stage, maybe just remove them all. If there actually is a serious user, maybe that pass could instead recompute them and also be able to trust them fully.

Also adding an assertion which is unrelated to kill flags, but it seems to make sense (according to liberal assertion policy), to verify that the preceding definition is in fact identical in clearKillsForDef(). This is a little border-line as this requires having an extra argument to that function, but my guess is that this will not matter. Could possibly wait with this as a a separate patch, or even keep skipping it.

Fixes #117783

@arsenm @nikic @RKSimon @jayfoad @topperc @uweigand

JonPsson1 committed:
Handle multiple kills for a preceding definition.

arsenm commented:
> If there actually is a serious user, maybe that pass could instead recompute them and also be able to trust them fully.

No, there should be no users. All users should be changed to not rely on kill flags

arsenm commented on llvm/lib/CodeGen/MachineLateInstrsCleanup.cpp:
```diff
@@ -113,33 +113,39 @@ bool MachineLateInstrsCleanup::runOnMachineFunction(MachineFunction &MF) {
+  // be ok to stop after removing that (and any other) kill-flag, but it
+  // doesn't seem noticeably faster while it would be a bit more complicated.
+  Reg2MIVecMap &MBBKills = RegKills[MBB->getNumber()];
+  if (MBBKills.contains(Reg))
+    for (auto *KillMI : MBBKills[Reg])
+      KillMI->clearRegisterKills(Reg, TRI);
```
Double map lookup

JonPsson1 replied:
Doesn't the [] operator add an entry to the map if its missing? In this case a mapping of Reg to an empty vector.

efriedma-quic replied:
Maybe:

```
  if (auto Kills = MBBKills.find(Reg); Kills != MBBKills.end())
    for (auto *KillMI : *Kills)
```

JonPsson1 replied:
ok - tried this now, although I think it's a bit less readable.

arsenm commented on llvm/test/CodeGen/SystemZ/machine-latecleanup-kills.mir:
```diff
@@ -0,0 +1,52 @@
+# RUN: llc -mtriple=s390x-linux-gnu -run-pass=machine-latecleanup %s -o - -mcpu=z16 \
```
Use update_mir_test_checks

JonPsson1 replied:
done

arsenm commented on llvm/lib/CodeGen/MachineLateInstrsCleanup.cpp:
```diff
@@ -225,8 +231,8 @@ bool MachineLateInstrsCleanup::processBlock(MachineBasicBlock *MBB) {
       } else if (MI.findRegisterUseOperandIdx(Reg, TRI, true /*isKill*/) != -1)
-        // Keep track of register kills.
-        MBBKills[Reg] = &MI;
+        // Keep track of all instructions that fully or partially kills Reg.
+        MBBKills[Reg].push_back(&MI);
```
You could also just do MRI.clearKillFlags any time it could matter

JonPsson1 replied:
That would be a lot simpler, just need to be sure it doesn't matter anywhere.

efriedma-quic replied:
It's fine.

JonPsson1 replied:
I think this has to be done per my previous experiments and comments from Dec 10 - there are many failures if I do this. It seems best to me to take care of the kill flags properly here at this moment until other passes are not dependent on them. (What I did on Dec 10 was to remove all kill flags that were affected by the pass, but now I am thinking maybe you meant to only remove those more rare kill flags of the super-register that this patch handles?)

JonPsson1 commented:
> > If there actually is a serious user, maybe that pass could instead recompute them and also be able to trust them fully.
> 
> No, there should be no users. All users should be changed to not rely on kill flags

There seems to be users in the Target backends - which could include late things like expandPostRAPseudo(), perhaps. There are also some use of isKill() in BranchFolding and probably other places. Do you think I could try to just remove kill flags a bit more carelessly ("ahead of time", even in the many cases where there is no following identical definition), and go ahead with this if no tests fail? Or are you saying there is some work left to be done in places that we should aim to do first?

JonPsson1 committed:
Updates per review.

JonPsson1 committed:
Go back to clearing of specific kill-flags.

JonPsson1 commented:
> You could also just do MRI.clearKillFlags any time it could matter

I tried this now, and found that
 - Still need to update live-in lists recursively: At first I just added Reg as live-in to the current MBB which the verifier accepted (it would have been better if it could report a missing definition for a live-out register in the predecessor). It however became clear that with the CFG optimizations running later on, all the predecessors up the CFG need to have the live-in list updated.
 - It's still a nice simplification to the algorithm (to take a look, see HEAD^ on the branch / dc4ff43).
 - Practically NFC on SPEC / SystemZ: only 7 files change without any real impact. This seemed due to BranchFolding.cpp mostly, and also MachineCopyPropagation in one case.
 - Only two CodeGen tests fail, however CodeGen/Thumb/frame-access.ll crashes with an assertion, which was unexpected:
 
```
MachineOperand::isKill() const: Assertion `isReg() && "Wrong MachineOperand accessor"'

bool ARMLoadStoreOpt::CombineMovBx(MachineBasicBlock &MBB) {
  MachineBasicBlock::iterator MBBI = MBB.getFirstTerminator();
  if (MBBI == MBB.begin() || MBBI == MBB.end() ||
      MBBI->getOpcode() != ARM::tBX_RET)
    return false;

  MachineBasicBlock::iterator Prev = MBBI;
  --Prev;
  if (Prev->getOpcode() != ARM::tMOVr ||
      !Prev->definesRegister(ARM::LR, /*TRI=*/nullptr))
    return false;

  for (auto Use : Prev->uses())
=>  if (Use.isKill()) {

(gdb) p Prev->dump()
  $lr = frame-destroy tMOVr $r1, 14, $noreg
 (gdb) p Use.dump()
14
```
It seems there is some assumption there that some register operand must be a kill.
- There is also ReachingDefAnalysis.cpp that checks for kills, which is used in BreakFalseDeps.cpp (only used by ARM / X86 at the moment).

Due to the above and in particular the crash in ARMLoadStoreOpt (which has a lot of checks for kills), I reverted back to the tracking of kill flags and clearing them carefully as per before. I would agree that there is room for improvement here, and at least on SystemZ it seems completely fine to remove any and all kill flags after regalloc.

JonPsson1 commented:
Ping!

JonPsson1 commented:
Ping!

Adding some more potential reviewers:
@vpykhtin 
@nickdesaulniers 
@efriedma-quic

JonPsson1 committed:
Updated per review

JonPsson1 commented:
updated per review

JonPsson1 commented:
Ping! Does this look ok now?

JonPsson1 commented:
Ping. Given that this was a problem reported by the MachineVerifier and that the patch is an improvement which handles cases of super-reg kills, I think I need to push this soon, if no one can see anything wrong with it.

The discussion around kill flags remains: I have tried your suggestions to simply remove them liberally, but that hasn't really worked. Tests breaks and even assertions trigger. So I think this is unfortunately a topic that would need a separate discussion: Should we remove the burden on optimizers to update kill flags, and if so how? Find lost optimizations acceptable, or recompute kill flags for passes that really depend on them?

@vpykhtin @nickdesaulniers

JonPsson1 commented:
Adding more reviewers with the hopes that someone can take a look at my handling of kill flags here, thanks.

@fhahn @MatzeB @jayfoad @topperc @bevin-hansson @serguei-katkov @qcolombet @preames @sdesmalen-arm @perlfu

perlfu commented on llvm/lib/CodeGen/MachineLateInstrsCleanup.cpp:
```diff
@@ -113,33 +113,39 @@ bool MachineLateInstrsCleanup::runOnMachineFunction(MachineFunction &MF) {
+  // Definition in current MBB: done.
+  Reg2MIMap &MBBDefs = RegDefs[MBB->getNumber()];
+  MachineInstr *DefMI = MBBDefs[Reg];
+  assert(DefMI->isIdenticalTo(*ToRemoveMI) && "Previous def not identical?");
```
Can you give more detail on the addition of this assert? Specifically because `ToRemoveMI` seems to exist only to support this?

JonPsson1 replied:
Yeah, I guess maybe that assertion could be a separate patch. The whole idea of the pass is to remove a redundant definition because there is an identical dominating one. Given the CFG traversal involved here (DefMI may be in a predecessor), it seems fair to do the assertion. Should I remove it from this patch now, you think?

perlfu reviewed (commented):
I think the substance of the change is good.  Sub/super register kills definitely need to be handled in this case.  This patch seems like a reasonable minimal change to address that need.

Can you add one further test that checks when the defs are super register and a kill is only for a sub register? I don't think existing tests cover this.

JonPsson1 committed:
Add test

JonPsson1 commented:
> Can you add one further test that checks when the defs are super register and a kill is only for a sub register? I don't think existing tests cover this.

done.

JonPsson1 commented:
@perlfu thanks for review!
